snuba/datasets/errors_replacer.py

```diff
@@ -667,15 +665,92 @@ def process_unmerge(
+
+    return LegacyReplacement(
+        count_query_template, insert_query_template, query_args, query_time_flags
+    )
+
+
+def _convert_hash(
+    hash: str, state_name: ReplacerState, convert_types: bool = False
+) -> str:
     if state_name == ReplacerState.ERRORS:
-        query_args["hashes"] = ", ".join(
-            ["'%s'" % str(uuid.UUID(_hashify(h))) for h in hashes]
-        )
+        if convert_types:
+            return "toUUID('%s')" % str(uuid.UUID(_hashify(hash)))
+        else:
+            return "'%s'" % str(uuid.UUID(_hashify(hash)))
     else:
-        query_args["hashes"] = ", ".join("'%s'" % _hashify(h) for h in hashes)
+        if convert_types:
+            return "toFixedString('%s', 32)" % _hashify(hash)
+        else:
+            return "'%s'" % _hashify(hash)
+
+
+def process_unmerge_hierarchical(
+    message: Mapping[str, Any],
+    all_columns: Sequence[FlattenedColumn],
+    state_name: ReplacerState,
+) -> Optional[Replacement]:
+    timestamp = datetime.strptime(message["datetime"], settings.PAYLOAD_DATETIME_FORMAT)
+    all_column_names = [c.escaped for c in all_columns]
+    select_columns = map(
+        lambda i: i if i != "group_id" else str(message["new_group_id"]),
+        all_column_names,
+    )
+
+    primary_hash = message["primary_hash"]
+    assert isinstance(primary_hash, str)
+
+    hierarchical_hash = message["hierarchical_hash"]
+    assert isinstance(hierarchical_hash, str)
+
+    where = """\
+        PREWHERE primary_hash = %(primary_hash)s
+        WHERE group_id = %(previous_group_id)s
+        AND has(hierarchical_hashes, %(hierarchical_hash)s)
+        AND project_id = %(project_id)s
+        AND received <= CAST('%(timestamp)s' AS DateTime)
+        AND NOT deleted
+    """
+
+    count_query_template = (
+        """\
+        SELECT count()
+        FROM %(table_name)s FINAL
+    """
+        + where
+    )
+
+    insert_query_template = (
+        """\
+        INSERT INTO %(table_name)s (%(all_columns)s)
+        SELECT %(select_columns)s
+        FROM %(table_name)s FINAL
+    """
+        + where
+    )
+
+    query_args = {
+        "all_columns": ", ".join(all_column_names),
+        "select_columns": ", ".join(select_columns),
+        "previous_group_id": message["previous_group_id"],
+        "project_id": message["project_id"],
+        "timestamp": timestamp.strftime(DATETIME_FORMAT),
+        "primary_hash": _convert_hash(primary_hash, state_name),
+        "hierarchical_hash": _convert_hash(
+            hierarchical_hash, state_name, convert_types=True
+        ),
+    }
 
+    # TODO: In principle Sentry replaces every single event within the group
+    # and assigns it a new group ID. For that it sends multiple of these
+    # replacement messages. We could get rid of NEEDS_FINAL in favor of
+    # EXCLUDE_GROUPS here if we had a way to run multiple statements as part of
+    # one replacements.
     query_time_flags = (NEEDS_FINAL, message["project_id"])
```

Comment:
So basically what happens here is that we are always going to send a group of replacement messages that will make the old group disappear. 
Under normal circumstances these replacements will be executed essentially shortly one after the other.
It seems that we are using FINAL (for up to 24 hours) to bridge a short gap of inconsistency (between the first replacement and the last replacement are executed).
I would expect the split operation to be a lot more common than an unmerge, so the imapct of FINAL on the project would be bigger (the first time you split a group on project 1 that whole project will become FINAL). 
Any chance we can accept this temporary inconsistency and either set EXCLUDE_GROUP on the first replacement or on the last ? Anyway the whole operation is not synchronous for the user. They trigger the split and, later, this happens. Event running evenrything in one query would introduce a period of time where data is essentially duplicated as the FINAL flag has not yet been applied.

Otherwise I would suggest to build the replacement in a single query asap to reduce the timeframe data is inconsistent. Running multiple queries in one replacement action would not make things substantially more consistent than multiple replacement messages as the flag would have to be set either before the first or after the last.

Reply:
>It seems that we are using FINAL (for up to 24 hours) to bridge a short gap of inconsistency (between the first replacement and the last replacement are executed).

I can accept temporary inconsistencies and I see the performance concern. I can replace `NEEDS_FINAL` with `None` here and either 1) do as you say and add an EXCLUDE_GROUP in the last replacement 2) send an additional `exclude_groups` message at the end (which already exists for reprocessing)

But now you've got me wondering if NEEDS_FINAL could be used to bridge over temporary inconsistencies everywhere: apply final to all queries while e.g. reprocessing is running (or really any process in Sentry which emits more than one replacement), stop applying when it is done and set EXCLUDE_GROUPS instead. Anyway, for another time.

Reply:
for now I've comitted option 2.